opencog/cython/opencog/value.pyx

```diff
@@ -1,60 +1,47 @@
-    """Factory method to construct Value from C++ ValuePtr (see
-    http://docs.cython.org/en/latest/src/userguide/extension_types.html#instantiation-from-existing-c-c-pointers
-    for example)"""
-    cdef Value proto_atom = Value.__new__(Value)
-    proto_atom.shared_ptr = shared_ptr
-    return proto_atom
-
-cdef list vector_of_doubles_to_list(const vector[double]* cpp_vector):
-    list = []
-    it = cpp_vector.const_begin()
-    while it != cpp_vector.const_end():
-        list.append(deref(it))
-        inc(it)
-    return list
-
-cdef list vector_of_strings_to_list(const vector[string]* cpp_vector):
-    list = []
-    it = cpp_vector.const_begin()
-    while it != cpp_vector.const_end():
-        list.append((<bytes>deref(it).c_str()).decode('UTF-8'))
-        inc(it)
-    return list
-
-cdef list vector_of_values_to_list(const vector[cValuePtr]* cpp_vector):
-    list = []
-    it = cpp_vector.const_begin()
-    cdef cValuePtr value
-    while it != cpp_vector.const_end():
-        value = deref(it)
-        if is_a(deref(value).get_type(), types.Value):
-            list.append(createProtoAtom(value))
-        else:
-            # TODO: Support Atoms as members of LinkValue requires inheriting
-            # Atom from Value and constructor to create Atom from cHandle.
-            raise TypeError('Only Values are supported '
-                            'as members of LinkValue')
-        inc(it)
-    return list
-
-cdef cValue* get_value_ptr(Value protoAtom):
-    """Return plain C++ Value pointer, raise AttributeError if
-    pointer is nullptr"""
-    cdef cValue* ptr = protoAtom.shared_ptr.get()
-    if ptr == NULL:
-        raise AttributeError('Value contains NULL reference')
-    else:
-        return ptr
+cdef class ValuePtr:
+    """C++ ValuePtr object wrapper for Python clients. Cython cannot create
+    Python object constructor which gets C++ pointer. This class is used to
+    wrap pointer and make it possible to initialize Value in usual
+    constructor (see
+    http://docs.cython.org/en/latest/src/userguide/extension_types.html#instantiation-from-existing-c-c-pointers)."""
+
+    @staticmethod
+    cdef ValuePtr create(cValuePtr shared_ptr):
+        """Factory method to construct ValuePtr from C++ cValuePtr"""
+        cdef ValuePtr value_ptr = ValuePtr.__new__(ValuePtr)
+        value_ptr.shared_ptr = shared_ptr
+        return value_ptr
 
 cdef class Value:
     """C++ Value object wrapper for Python clients"""
 
+    @staticmethod
+    cdef Value create(cValuePtr shared_ptr):
+        """Factory method to construct Value from C++ cValuePtr using ValuePtr
+        instance."""
+        return Value(ValuePtr.create(shared_ptr))
+
+    def __init__(self, value_ptr):
+        self.value_ptr = value_ptr
+
+    cdef cValuePtr get_c_value_ptr(self):
+        """Return C++ ValuePtr instance"""
+        return self.value_ptr.shared_ptr
+
+    cdef cValue* get_c_value(self):
```

Comment:
Names are confusing: get_c_value_ptr returns shared pointer and get_c_value returns raw pointer, which is pointer type too. Clear names are get_shared_ptr and get_raw_ptr. But it seems shared pointer should be enough for most cases.

Reply:
or just get_ptr and get_raw_ptr

Reply:
I have removed `get_c_value` method and replaced it by `get_c_value_ptr` calls